### ISSUE TITLE
Scrooge no longer can compile thrift-only JARs

### ISSUE THREAD

**Issue by vsabella** [twitter/scrooge#199]
The change for Scrooge to allow thrift files to be published in dependency jars, 1ab8b6650dddbf54f5d2bfa81a9321975ecfca3c appears to prevent thrift-only jars from being an option now.

Our team ships thrift-only JARs which we expect Scrooge to compile but as of 3.20 this is no longer the case.

**Comment by nshkrob** [comment]
@jedesah FIY. I guess we should handle thrift-only jars separately.

**Comment by jedesah** [comment]
@nshkrob That sounds reasonable to me.

**Comment by jstultz** [comment]
Just wanted to chime in with some support for a change to consider thrift-only jars. I'm in the process of updating scrooge in our infrastructure, and the version we are currently using predates the referenced commit. Internally, for thrift dependencies, we exclusively use jars containing only thrift sources. For now, we are going to work around the issue with the following setting:

```scala
  val compileThriftDepsSetting = scroogeThriftSources in Compile ++= {
    (scroogeUnpackDeps in Compile).value.flatMap { dir => (dir ** "*.thrift").get }
  }
```

However, we've generally found working with jars containing only thrift sources to be preferable to packaging compiled classes, and it would be nice to have first-class support from scrooge for this (we similarly have our own artifact task(s) to package up just the thrift files and not the compiled classes, though they do perform a compile to ensure that the thrift is valid and we don't publish thrift source that won't compile).

It's clear that in general, the thrift source is often required, for exactly the reasons that https://github.com/twitter/scrooge/pull/164 states. However, it's not clear to me that the compiled classes are ever really necessary when the thrift sources are available. Packaging/publishing compiled classes potentially bakes in a dependency on a specific version of scrooge (Either because the compiled output is different in an incompatible way between versions, or because the scrooge runtime classes are different. See https://github.com/twitter/scrooge/issues/190 for one such example).

We use thrift and scrooge extensively for our finagle service interfaces; it seems unfortunate to impose requirements on the respective versions of scrooge used by a scala client using scrooge and a scala server using scrooge. Why should it matter, as long as the thrift protocol version is the same? It's perfectly reasonable, after all, for clients to not be using scrooge at all (as we do for our ruby and go clients)

I've seen references to issues arising from having to recompile the thrift definitions many times in all projects that use it (for example: https://github.com/twitter/scrooge/commit/1ab8b6650dddbf54f5d2bfa81a9321975ecfca3c#commitcomment-12432540), but haven't seen specifics on what those issues are. Obviously at some level it is redundant and unnecessary extra work, but the cost of compilation isn't that high. Are there other concerns I am unaware of and haven't thought of? From my perspective at this point the tradeoffs are clearly in favor of only packaging and publishing thrift sources (except, of course, all the way downstream when it is actually used in a project at runtime). For example, if we had published compiled classes for all of our service interfaces along with a dependency on scrooge 3.15, the upgrade to scrooge 4.3.0 that I'm working on right now would be much more difficult (having to orchestrate updates in order of service dependencies, etc).

Perhaps it would work nicely to have scrooge support publishing either a thrift-only JAR, or a compiled-classes-only JAR, or both (but separately), and have it support dependencies of either type. If there are thrift definitions present, add them to the thrift include path and also compile them; if there are compiled classes present, take them as is and add them to the classpath, etc. Using the thrift source of a dependency as an include to compile thrift code, and also using precompiled classes from that same thrift source, with no guarantee that the same version of scrooge is used, seems inherently problematic to me.

**Comment by leadVisionary** [comment]
A JAR file is a Java Archive File. 
Apache .thrift files are by definition an Interface Definition Language, which is platform independent. 

Packaging thrift files in JARs doesn't make sense. Should Scrooge support Zip? GZip? RAR?

The only things that should be published as a JAR are the compiled Java/Scala/JVM-lang classes that the compiler generates.

**Comment by jedesah** [comment]
@jstultz Issues arise from attempting to create "fat jars" where multiple versions of the same class exist in the classpath. Surprisingly, HotSpot is fine with it, the problem only arises (or arose at some point in the past, thankfully I don't have to work with thrift anymore these days) when using a tool such as `onejar` to create a "fat jar". I think it's totally reasonable to have both options as an organization such as yours might care much more about Scrooge version independance than that issue. At the time, my organization needed to use `onejar` for deployment and the existence of dozens of identical classes was a serious problem whereas syncing versions of Scrooge was easy enough for us.

@leadVisionary 

> The only things that should be published as a JAR are the compiled Java/Scala/JVM-lang classes that the compiler generates.

Resources (such as image files, data files, etc)  are commonly packaged in jar files and java has always had support for this through APIs in the standard library, allowing one to safely include necessary assets in a jar file that the executable requires in order to run properly in a platform independent manner.

But maybe that's not what you meant, maybe you meant that not having any java classes is a perversive use of jars. There may be some truth to that, but there is so much "relatively" good dependency management tools surrounding the java ecosystem that are somewhat hardcoded to understand `jar` files that it's rather practical to use them to send thrift files that are intended to be used in the java ecosystem.

**Comment by mosesn** [comment]
It seems annoying to me to have to pack the scrooge generator and do generation at startup, but if that style works well for you, more power to you!  Would you like to make a PR?

**Comment by jstultz** [comment]
@leadVisionary : as @jedesah pointed out, there is plenty of precedent for resources other than compiled JVM bytecode being packaged into JARs; configuration files and various other resources. WebJars is an example of a use of JARs that do not contain any JVM byte code at all. Whether or not the files themselves contain compiled Java classes, they're still used in a Java environment which is built around JARs.

@jedesah : we have also had a number of issues with fat jars (using sbt-assembly, though). We are generally moving away from that for most of our services, but for those projects that we are still building fat jars for, the strategy of publishing thrift-only artifacts actually works quite well; when the fat jar is actually built for an application to be deployed, the build pulls in the thrift files for the thrift dependencies. If a particular project occurs in the dependency twice (presumably due to transitive dependencies), then there's not an issue with conflicting class files because they haven't been built yet; the thrift files are compiled into scala and then to class files by the ultimate downstream project.

@mosesn : I may be misunderstanding your comment, or which "startup" you're referring to, but the way we have things set up, we're not packing a scrooge generator at all. We have an internal SBT plugin with project helpers for creating and depending on projects that simply define a thrift interface. When the thrift interface projects are published, we build a jar that just contains the thrift definitions and publish it in the thrift configuration.

Then, when pulling in a thrift project as a dependency, we depend on the thrift configuration of that project. Our SBT plugin pulls in the scrooge SBT plugin as a dependency, so any project that depends on a thrift interface will compile it happily, even if it's using a different version than other projects using the same thrift interface.

This does mean as noted that in theory the thrift definition is compiled once for every project that depends on it, directly or transitively. In practice, this hasn't been an issue. Certainly this overall pattern is not one-size-fits-all, but it works for us, and it's nice to have the option to do it this way.

I'd be happy to make a PR once I've got a bit of time; before I do so, I'd just like to clarify what it would consist of. Does what I outlined in my previous comment seem reasonable?

> Perhaps it would work nicely to have scrooge support publishing either a thrift-only JAR, or a compiled-classes-only JAR, or both (but separately), and have it support dependencies of either type. If there are thrift definitions present, add them to the thrift include path and also compile them; if there are compiled classes present, take them as is and add them to the classpath, etc. Using the thrift source of a dependency as an include to compile thrift code, and also using precompiled classes from that same thrift source, with no guarantee that the same version of scrooge is used, seems inherently problematic to me.

**Comment by mosesn** [comment]
@jstultz so when you run in production you run it with sbt?

**Comment by leadVisionary** [comment]
I'm fully aware that resources can be packages in JARs, thank you for the education. Precedent is not the same as original design intent. If you find yourself making hacks on hacks on hacks on hacks, you should probably take a step back and ask yourself about the design of your system.

Do you include .java files in your JARs? Generally, not on your binary distributions. You don't do so because you rely on javac to compile them to a .class format that is readable by any JVM that adheres to spec. Consequently, including them pollutes archive size and creates potential security vulnerabilities.

A sane Thrift workflow would use Scrooge to compile your Thrift, generate your classes, and distribute that to another artifact repository that downstream applications can take dependency on.

The missing feature here is that you want to be able to reference the outputs of other models in your Thrift IDL without having to recompile it. `include` semantics currently force you to have `foo.thrift` available for compilation. If Scrooge were able to lookup known classes on the classpath when encountering `include` and skip already compiled classes, you could package your models nicely and have dependencies without weird "thrift-only" packages.

**Comment by jstultz** [comment]
@mosesn oh, no, definitely not. I'll give a simplified example.

Suppose we have two finagle services serving thrift interfaces, `foo` and `bar`, where `bar` is a client of `foo`. Each has it's own git repository and associated SBT configuration, within which there are two projects, `foo-interface`, `foo-server` (same for `bar`).

`foo-interface` contains only thrift definition. The published artifact for this project is a jar containing the thrift files (though the publish task does invoke scrooge to compile the thrift to ensure that it is valid).

`foo-server` has a dependency on `foo-interface`. For deployment, we build a fat jar, and part of building that fat jar is using scrooge to compile the thrift to scala, and then all of the scala code is then compiled to class files and combined with dependencies, etc. That fat jar is then used in production. (At least, in the fat jar scenario that we're moving away from; more commonly now we'll build a docker image containing dependency jars, including a fully compiled jar for the `foo-interface` project).

`bar-interface` is just like `foo-interface` here, and then `bar-server` is much like `foo-server`, except that it depends on both `foo-interface` and `bar-interface`; when built for deployment it will have fully compiled versions of both.

Another case is where we have thrift interfaces that depend on other thrift interfaces (typically where we have some structs that are common between multiple service interfaces). Here, imagine that `foo-interface` depends on `shared-interface` (where there is no `shared-server`, it's just a shared thrift definition). `shared-interface` gets published as  a jar containing thrift files, `foo-interface` still gets published as a jar containing thrift files, but with a dependency on `shared-interface`. When `foo-server` gets built, it will pull in both `foo-interface` as well as `shared-interface`, and fully compile them for deployment.

The idea here for us is that the coupling between `foo-server` and `bar-server` is just via the thrift interface; there's really no reason why they need to be using versions of `foo-interface` built with the same version of scrooge, or even built with the same thrift compiler, or even running on the same platform/language.

Again, this works for us, but I understand if there's not agreement that this is a methodology that scrooge wants to target/support. If it is, I'm happy to contribute to make it work.

**Comment by mosesn** [comment]
Seems reasonable to me.  I'd definitely look at a PR!

**Comment by mosesn** [comment]
@jstultz are you still interested in tackling this by any chance?

**Comment by jstultz** [comment]
I could try to find some time if there were some broader interest in the change; we have a solution (mentioned above) that _mostly_ works for us now, but there are still some problems that I'd want to come up with a solution to before putting something upstream. 

Everything works fine for us in a simplified (unrealistic) world where all SBT subprojects are either thrift projects that publish thrift-only JARs, or service implementation projects that are just deployed directly (e.g. by building and publishing a docker image), but do not publish library JARs used by other projects.

Problems can arise when publishing scala library JARs that depend on thrift JARs: in our current build implementation, if library A depends on thrift project T, we publish T as a thrift only JAR, and then library A's build compiles T's thrift, and those compiled classes are then included in A's JAR. This is weird, but is currently not easy to avoid, and it works fine in this simplified scenario.

When problems can occur (along the lines of the fat JAR issue mentioned earlier by others in this issue) is when different libraries depend on the same thrift definitions; say now we have A -> T, B -> T, and C which depends on both A and B. A and B's JARs will both have compiled classes for T's definitions, causing conflicts that must be resolved in C.

Two options as I see it:
 - give up on thrift only JARs; this issue clearly demonstrates that it doesn't actually entirely avoid the scrooge version dependency coupling that I discussed earlier
 - configure other projects to not include compiled classes originating from thrift definitions in their own output, instead requiring all downstream dependent projects to compile them as necessary, not just those directly depending on the thrift projects as we do it now.

I don't love the latter option: we haven't attempted it in part because scrooge-sbt doesn't currently expose an sbt task referencing the generated files (so that they can be easily excluded from package output). This could be added to scrooge-sbt, but the problem with scrooge providing this dependency style is that it requires a lot of changes to how non-thrift projects are built. We are okay with this internally, but it's not clear to me that it is reasonable for it to be built-in functionality of scrooge-sbt.

**Comment by mosesn** [comment]
@jstultz thanks for the write-up, that makes a lot of sense.  What do you think is the right way forward on this issue?  Should we close it?  Is there something simple that we could add to scrooge that would make your life easier?

**Comment by jstultz** [comment]
The only idea I've got to resolve that issue is to make it easy to exclude compiled classes originating from the thrift projects from the `packageBin` output by providing some setting/task that yields the files as a value. If those files are known, it wouldn't be difficult, but it doesn't seem trivial: all scrooge-sbt does is put the generated scala files into managed sources, right? Determining the class files generated from the scala files in SBT is something I don't know how to do.

I'd _personally_ be fine with the issue being closed; I've come to the conclusion that distributing thrift-only JARs was the wrong approach. @liamstewart may have more thoughts to add; we've discussed possible solutions in our environment some.
